File: js/problems/2-add-two-numbers.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function (l1, l2) {
  let node = new ListNode();
  let res = node;
  let carry = 0;

  while (l1 || l2 || carry) {
      let num = 0;

      if (carry > 0) {
          num = num + carry;
          carry = 0;
      }

      if (l1) {
          num += l1.val;
          l1 = l1.next;
      }
      if (l2) {
          num += l2.val;
          l2 = l2.next;
      }

      if (num > 9) carry = 1;

      node.next = new ListNode(num % 10);
      node = node.next;
  }

  return res.next;
};

File: .leetcode/2.add-two-numbers.cpp
/*
 * @lc app=leetcode id=2 lang=cpp
 *
 * [2] Add Two Numbers
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
 public:
  ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
    /*
      Approach 1: None
      time complexity: O(n)
      space complexity: O(1)
    */

    ListNode* dummy = new ListNode(0);
    ListNode* current = dummy;
    int carry = 0;

    while (l1 != nullptr || l2 != nullptr || carry != 0) {
      int sum = carry;

      if (l1 != nullptr) {
        sum += l1->val;
        l1 = l1->next;
      }

      if (l2 != nullptr) {
        sum += l2->val;
        l2 = l2->next;
      }

      carry = sum / 10;
      current->next = new ListNode(sum % 10);
      current = current->next;
    }

    return dummy->next;
  }
};
// @lc code=end